e2e/tests-dfx/pull.bash

```diff
@@ -95,18 +97,153 @@ WARN: \`dfx:deps\` metadata not found in canister rrkah-fqaaa-aaaaa-aaaaq-cai."
 
     cd ../app
     assert_command_fail dfx pull
-    assert_contains "Failed while fetch and parse \`dfx:deps\` metadata from canister rrkah-fqaaa-aaaaa-aaaaq-cai."
+    assert_contains "Failed to fetch and parse \`dfx:deps\` metadata from canister rrkah-fqaaa-aaaaa-aaaaq-cai."
     assert_contains "Canister rrkah-fqaaa-aaaaa-aaaaq-cai not found."
 
     # 3. sad path: if dependency metadata cannot be read (wrong format)
     cd ../onchain
     cd src/onchain_b
-    ic-wasm empty.wasm -o main.wasm metadata "dfx:deps" -d "rrkah-fqaaa-aaaaa-aaaaq-cai;onchain_a" -v public
+    ic-wasm main.wasm -o main.wasm metadata "dfx:deps" -d "rrkah-fqaaa-aaaaa-aaaaq-cai;onchain_a" -v public
     cd ../../ # go back to root of "onchain" project
     dfx deploy
 
     cd ../app
     assert_command_fail dfx pull
-    assert_contains "Failed while fetch and parse \`dfx:deps\` metadata from canister ryjl3-tyaaa-aaaaa-aaaba-cai."
+    assert_contains "Failed to fetch and parse \`dfx:deps\` metadata from canister ryjl3-tyaaa-aaaaa-aaaba-cai."
     assert_contains "Failed to parse \`dfx:deps\` entry: rrkah-fqaaa-aaaaa-aaaaq-cai. Expected \`name:Principal\`."
 }
+
+@test "dfx pull can download wasm" {
+    # When ran with ic-ref, got following error:
+    # Certificate is not authorized to respond to queries for this canister. While developing: Did you forget to set effective_canister_id?
+    [ "$USE_IC_REF" ] && skip "skipped for ic-ref"
+    use_test_specific_cache_root # dfx pull will download files to cache
+
+    WASM_CACHE="$DFX_CACHE_ROOT/.cache/dfinity/wasms/"
+
+    assert_file_not_exists "$WASM_CACHE/ryjl3-tyaaa-aaaaa-aaaba-cai/canister.wasm"
+    assert_file_not_exists "$WASM_CACHE/rrkah-fqaaa-aaaaa-aaaaq-cai/canister.wasm"
+    assert_file_not_exists "$WASM_CACHE/r7inp-6aaaa-aaaaa-aaabq-cai/canister.wasm"
+
+    # system-wide local replica
+    dfx_start
+
+    install_asset pullable
+
+    # start a webserver to host wasm files
+    mkdir www
+    start_webserver --directory www
+
+    # prepare "onchain" canisters
+    cd onchain
+
+    echo -n -e \\x00asm\\x01\\x00\\x00\\x00 > src/onchain_a/main.wasm
+    ic-wasm src/onchain_a/main.wasm -o src/onchain_a/main.wasm metadata "dfx:wasm_url" -d "http://localhost:$E2E_WEB_SERVER_PORT/a.wasm" -v public
+
+    echo -n -e \\x00asm\\x01\\x00\\x00\\x00 > src/onchain_b/main.wasm
+    ic-wasm src/onchain_b/main.wasm -o src/onchain_b/main.wasm metadata "dfx:wasm_url" -d "http://localhost:$E2E_WEB_SERVER_PORT/b.wasm" -v public
+    ic-wasm src/onchain_b/main.wasm -o src/onchain_b/main.wasm metadata "dfx:deps" -d "onchain_a:rrkah-fqaaa-aaaaa-aaaaq-cai;" -v public
```

Comment:
Would it make sense to `dfx deploy` and `dfx canister create --all`, then use `dfx canister id` to obtain canister IDs, rather than hardcoding these canister ids?

Reply:
I overlooked that it's possible to get canister id before deploy it. I will fix this in another PR later.